### ISSUE TITLE
Hypercube interface

### ISSUE THREAD

**Issue by cscherrer** [cscherrer/TupleVectors.jl#14]
Let's start with Sobol sequences, only because we currently have code for these. If we're in `d` dimensions, the Sobol sequence is a sequence of length-d vectors. 

Actually, let's make this even more concrete. Suppose we want the low-discrepancy equivalent of
```julia
julia> using Distributions

julia> x = rand(MvNormal(ones(2)))
2-element Vector{Float64}:
 0.13812978810308807
 1.3544074368443249

julia> y = rand()
0.832947259863168

julia> nt = (;x,y)
(x = [0.13812978810308807, 1.3544074368443249], y = 0.832947259863168)
```

We can express this in terms of three "primitive" calls to `rand` (using inverse transform sampling for the Gaussian). So here we'd need something like
```julia
julia> using Sobol

julia> SobolSeq(3)
3-dimensional Sobol sequence on [0,1]^3
```

The way we currently handle this is with
```julia
struct SobolHypercube{k} <: Hypercube{k}
    seq :: SobolSeq{k}
    value :: Vector{Float64}
    index :: Ref{Int}  # start at zero

    function SobolHypercube(k::Int)
        seq = SobolSeq(k)
        value = Sobol.next!(seq)
        return new{k}(seq, value, 0)
    end
end
```

Here `value` is the current "raw sample" (element of the unit hypercube), and `index` is a pointer into it. But [`Ref`s are slow](https://github.com/JuliaLang/julia/issues/40369), so we should find a different way (and confirm the speed difference).

In a Zulip discussion about this, @mschauer pointed out [Stateful iterators](https://docs.julialang.org/en/v1/base/iterators/#Base.Iterators.Stateful). As Moritz pointed out in Slack, connecting to the iterator protocol would make the same interface available for lots of different cases

**Comment by cscherrer** [comment]
By "Let's start with Sobol sequences", I don't mean for new work, but for the discussion. Concrete code can help as a reference point, and might make the discussion easier.

**Comment by cscherrer** [comment]
By "Let's start with Sobol sequences", I don't mean for new work, but for the discussion. Concrete code can help as a reference point, and might make the discussion easier.

**Comment by ParadaCarleton** [comment]
For univariate distributions, sampling should be pretty easy. Every univariate distribution already has a quantile function implemented, which can be used to generate quasi-random samples from a quasi-random sequence on the d-dimensional uniform.

For multivariate distributions, things get a bit trickier; is there some already-implemented way to get the quantile of the `n`th component of a multivariate distribution, conditional on the first `n-1` components? If so, we can sample iteratively -- generate the first value using the inverse-cdf transform, then use the quantile function of the second component conditional on the first, and so on.

**Comment by cscherrer** [comment]
Inverse transform sampling is a good way to go for many cases, at least as a fall-back for univariate distributions. But if we have, say, a multivariate Gaussian, it's much better to draw samples from iid standard normal and then multiply and/or add to get to what you need.

There's a nice overview of this sort of "one-liner" at http://blog.shakirm.com/2015/10/machine-learning-trick-of-the-day-4-reparameterisation-tricks/

**Comment by ParadaCarleton** [comment]
> Inverse transform sampling is a good way to go for many cases, at least as a fall-back for univariate distributions. But if we have, say, a multivariate Gaussian, it's much better to draw samples from iid standard normal and then multiply and/or add to get to what you need.
> 
> There's a nice overview of this sort of "one-liner" at http://blog.shakirm.com/2015/10/machine-learning-trick-of-the-day-4-reparameterisation-tricks/

Yep, so multiplying by a proper scale matrix gives you multivariate Gaussians pretty easily, but it's not obvious how to generalize this. What if I need to preserve a low-discrepancy sequence while sampling from an inverse-Wishart, for instance?

**Comment by cscherrer** [comment]
I don't think there's one approach that will be best across the board. 

The question of preserving the low discrepancy is an interesting one. I had been thinking we're probably ok as long as
- We use a fixed number of calls (no rejection sampling)
- A given coordinate in the sequence plays the same role across samples
- Transformations are "reasonably smooth" (no space-filling curves)

So for example, for a Gaussian, Box-Muller seems fine, but I'd avoid Ziggurat sampling

**Comment by ParadaCarleton** [comment]
> I don't think there's one approach that will be best across the board.
> 
> The question of preserving the low discrepancy is an interesting one. I had been thinking we're probably ok as long as
> 
> * We use a fixed number of calls (no rejection sampling)
> * A given coordinate in the sequence plays the same role across samples
> * Transformations are "reasonably smooth" (no space-filling curves)
> 
> So for example, for a Gaussian, Box-Muller seems fine, but I'd avoid Ziggurat sampling

A visual comparison of Box-Muller samples:
![image](https://user-images.githubusercontent.com/71727937/119302242-a9dcf800-bc18-11eb-8333-b48f5066146c.png)

Against inverse transform samples:
![image](https://user-images.githubusercontent.com/71727937/119302312-c2e5a900-bc18-11eb-9831-6a45e742bbcc.png)

Both look decent to me, but neither looks optimal, interestingly enough. The top has spirals and the bottom looks a bit boxy. The second looks like if it's optimized on the taxicab metric, which kind of makes sense, while the top looks like it's optimized for some kind of weird metric based on the radius and the angle (again makes sense). My main concern is this might get worse in higher dimensions.

**Comment by mschauer** [comment]
This looks kind of good:

```
using Distributions, GoldenSequences
gauss(x) = (quantile(Normal(), x[1]), quantile(Normal(), x[2]))  
boxmueller(x) = (sqrt(-2log(x[1]))*cos(2pi*x[2]), sqrt(-2log(x[1]))*sin(2pi*x[2]))
x = map(boxmueller, take(GoldenSequence(2), 100))
p = scatter(x)
save("boxmueller.png", p)
```
![boxmueller](https://user-images.githubusercontent.com/1923437/119315400-14705100-bc76-11eb-94e2-a82212848374.png)

**Comment by mschauer** [comment]
Correction: It's about the same.

**Comment by cscherrer** [comment]
Could it be a limitation of GoldenSequences? Sobol seems to do fine:
```julia
using TupleVectors
using TupleVectors: unwrap
using Sobol

ω = SobolHypercube(2)

boxmuller = @with (;ω) 1000 begin
    z1 = rand(ω)
    z2 = rand(ω)
    (x = sqrt(-2log(z1))*cos(2pi*z2), y = sqrt(-2log(z1))*sin(2pi*z2))
end

using Plots

@with unwrap(boxmuller) begin
    scatter(x,y, legend=false)
end
```

![image](https://user-images.githubusercontent.com/1184449/119351804-c0ee0b00-bc55-11eb-94fa-292fe011b727.png)

**Comment by cscherrer** [comment]
Anyway, Box-Muller is at least interesting as a test case, to see how multivariate transforms might affect discrepancy. There's a question of whether Box-Muller itself makes sense at all, since we can do inverse transform sampling with `norminvcdf`.

I wonder if what we're seeing could have to do with the Jacobian of the transform. This didn't simplify all the way, but it's a start:
```julia
julia> using Symbolics, StatsFuns, LinearAlgebra

julia> @variables x y
2-element Vector{Num}:
 x
 y

julia> dx = Differential(x); dy = Differential(y);

julia> J(z) = expand(det(Symbolics.jacobian(z, [x,y])))
J (generic function with 1 method)

julia> J([sqrt(-2log(x))*cospi(2y), sqrt(-2log(x))*sinpi(2y)])          # Box-Muller
-6.283185307179586(x^-1)*(cospi(2y)^2) - (6.283185307179586(x^-1)*(sinpi(2y)^2))

julia> J([norminvcdf(x), norminvcdf(y)])                     # Inverse transform sampling
6.283185307179587exp(SpecialFunctions.erfcinv(2x)^2)*exp(SpecialFunctions.erfcinv(2y)^2)
```

**Comment by cscherrer** [comment]
Interesting, here's GoldenSequences with inverse transform sampling:
![image](https://user-images.githubusercontent.com/1184449/119372338-087f9180-bc6c-11eb-92b5-e4b6978f4bb5.png)

**Comment by cscherrer** [comment]
I think we're getting at some different questions here:
1. When a user is deciding between Sobol and GoldenSequences, what should they base that decision on? (should be an issue in GoldenSequences repo)
2. What transformations should be used by default? Should others be available, maybe through keyword arguments? (should be a different issue here)

For this issue, let's focus on the question of the random interface, especially making sure it has good performance and predictable results.

There's a file `src/optional/MeasureTheory.jl` in this repo that's just this:
```julia
using .MeasureTheory
using ..TupleVectors

function Base.rand(ω::Hypercube, d::ParameterizedMeasure)
    Dists.quantile(distproxy(d), rand(ω))
end

function Base.rand(ω::Hypercube, d::ProductMeasure)
    [rand(ω, dj) for dj in d.data]
end
```

More on this in a bit...
